jsonpb/jsonpb.go

```diff
@@ -640,7 +644,15 @@ func (u *Unmarshaler) unmarshalValue(target reflect.Value, inputValue json.RawMe
 	// Allocate memory for pointer fields.
 	if targetType.Kind() == reflect.Ptr {
 		target.Set(reflect.New(targetType.Elem()))
-		return u.unmarshalValue(target.Elem(), inputValue, prop)
+		err := u.unmarshalValue(target.Elem(), inputValue, prop)
+
+		// Set to a zero (nil pointer) if wrapper type is nulled
+		if err == nullWrapperError {
```

Comment:
Instead of using an error for undoing the construction and setting to nil, it would be better to do it ahead.  See the diff I've attached.